src/main/java/liquibase/ext/spanner/SpannerCreateTableGenerator.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * <p>https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * <p>Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package liquibase.ext.spanner;
+
+import liquibase.database.Database;
+import liquibase.sql.Sql;
+import liquibase.sql.UnparsedSql;
+import liquibase.sqlgenerator.SqlGeneratorChain;
+import liquibase.sqlgenerator.core.CreateTableGenerator;
+import liquibase.statement.core.CreateTableStatement;
+import liquibase.structure.DatabaseObject;
+import liquibase.util.StringUtils;
+
+public class SpannerCreateTableGenerator extends CreateTableGenerator {
+
+  public SpannerCreateTableGenerator() {}
+
+  @Override
+  public Sql[] generateSql(
+      CreateTableStatement statement, Database database, SqlGeneratorChain sqlGeneratorChain) {
+    Sql[] res = super.generateSql(statement, database, sqlGeneratorChain);
+    // Move the PRIMARY KEY statement from inside the table creation to outside.
+    StringBuilder buffer = new StringBuilder(", PRIMARY KEY (");
+    buffer.append(
+        database.escapeColumnNameList(
+            StringUtils.join(statement.getPrimaryKeyConstraint().getColumns(), ", ")));
+    buffer.append(")");
+
+    String pk = buffer.toString();
+    String sql = res[0].toSql();
+    sql = sql.replace(pk, "");
+    // Append PRIMARY KEY (without the leading ,)
```

Comment:
You're re-generating the PRIMARY KEY part, stripping it, and then adding it back.
Do we validate if there's any primary key? Spanner can't handle this, but Liquibase is happy creating tables without it.

Reply:
Good point, I've added a validation step for that.

Cloud Spanner requires a Primary Key to be defined, but the list of columns is allowed to be empty. That creates a table that may hold at most 1 row. It is however not possible to define such a table using the meta model of Liquibase.